### DISCUSSION TITLE
Maintenance mode

### DISCUSSION THREAD

**Discussion by 0x4E4448** [dani-garcia/vaultwarden#3207] [Ideas]
I'm running Vaultwarden in Kubernetes with an SQLite database. My plan is to eventually move to MariaDB. In this configuration, I cannot stop the running Vaultwarden container to backup the files.

I suggest a flag in the configuration file that can be flipped from 0 to 1 to enable a maintenance mode, which will stop web vault logins, prevent plugins from syncing, write any pending data to the database and, if possible, lock the database to prevent further changes. This would allow for backing up the database and related files without risking corruption and inconsistency, especially in larger instances. The reverse should happen when flipping from 1 to 0.

At very least, it should prevent logins and syncing.

**Comment by BlackDex** [comment]
Or, just use the right commands https://github.com/dani-garcia/vaultwarden/wiki/Backing-up-your-vault#sqlite-database-files

Or create a small script to login into the admin and call the backup endpoint which calls the `VACUUM into` command.

**Reply by 0x4E4448** [reply]
I'm aware of the commands, but that doesn't prevent changes to the database and inconsistencies between the database and files. The tools to do also may not be available to the backup container without additional customization, which shouldn't be necessary. The backup container should be able to backup without having interact with the database itself.

**Reply by BlackDex** [reply]
Well, I'm not seeing this a something useful to be added my self actually.
The only thing we could do is block every request, which is the same is either stopping the container, or prevent the reverse proxy from passing on the requests.

Running that `VACUUM into` command doesn't make stuff inconsistent.
If you use the `admin` interface to trigger the backup command, your backup container doesn't have to interact with the database it self.

Also, having such a maintenance option doesn't prevent inconsistencies between database and files. For file uploads, they first mention that a file is going to be uploaded, after that the file is going to be uploaded using information from the response. If you enable the maintenance mode right after it requested the info, but the file wasn't uploaded yet, that would break.

So, i still say, use the admin API to trigger a backup, and backup all the files.

**Reply by 0x4E4448** [reply]
I see it as very useful for multiple reasons. For one, all software should have a flag that allows an admin to prevent logins while work is performed without having to take down the entire service because some actions may require the service to be running. It's as simple as stopping the logins and presenting a maintenance page in many cases.

Here I'm dealing with Kubernetes, which handles things a bit differently than Docker. You can't simply stop a container in Kubernetes. You can cause a container to restart or you can stop the entire pod, but that stops your backup container and deletes your data, unless you have persistent storage. Persistent storage isn't the right solution here because the quantity of data isn't significant.

The way it generally works in Kubernetes is for an initcontainer to download the files from S3 or some other storage service and place them into an emptyDir, which is just an empty volume. Once that has completed, that initcontainer will stop and the service container will start, as well as a second container for backups that runs in parallel. The service container is, in this case, just the Vaultwarden container and it has access to the emptyDir that contains the SQLite database and the attachments. The second container runs a cronjob periodically that backs up that data.

The emptyDir with the data exists while the pod (the set of containers) are running. If a single container fails, it restarts and continues using the same data in the emptyDir. If the host fails and the pod restarts on the same or a different host, then this whole process starts over using the most recent backup.

The simplest way to do this is to backup the entire data_dir without having to export everything *because* exporting also means you have to import everything back into the database, which doesn't exist before Vaultwarden starts. *THIS* is why it is best for there to be a maintenance flag. It prevents backing up a database that is being written to.

It's a little different when using MariaDB and I suspect different when using Postgres, but suffers similar issues regarding corruption when the files are backed up while the database container is still in use. Preventing login and synchronization would reduce the risk of the database being written to at the time.

**Comment by BlackDex** [comment]
It's going to be difficult to not have people login, but also not to be forced to logout i think. But I'm not sure how the clients would handle an arbitrary maintenance message actually.

Second, i still do not see the issue of using the `VACUUM into` backup API endpoint provided via the admin. It will create a new file with the data present at that specific point in time. That file will not be used by any thing and thus is free from corruption at that point. Yes you would need to rename that if you need to restore that specific version but it is at least a valid backup.

Also it is not a good idea to just backup the database while it's open in regards to sqlite. While it may not be written to, it could still have some data in mem or in the shm or wal files not fully processed. That could also mean corruption. The only safe way is using those commands to backup. Same btw for MySQL or PostgreSQL, you need to dump the database or stop the database first and then make a backup.

And not using a persistent volume is a choice, k8s works just fine using persistent volumes. And in regards to not being able to import (with which i think you mean copy over all the data from S3 to the data folder), that is something you can script and put in the `vaultwarden.d` directory which gets executed before Vaultwarden starts. And for that you could use configMaps to create and update that script.

**Reply by 0x4E4448** [reply]
I’m just addressing your last comment as I can’t fully respond at the moment.

Yes, persistent volumes work fine with Kubernetes but are generally avoided because having multiple hosts for the Vaultwarden container to run means you have to have multiple persistent volumes, one for each node, and you have to keep those in sync somehow, so you run into the same issue with copying a running database. Sure, an NFS share could be used but then you’re back down to a single point of failure. Using the method I use for everything, so far, provides more redundancy because the container can come up on any node without dealing with multiple volumes and keeping them in sync.

**Reply by BlackDex** [reply]
You are still having a single point of failure when not using a persistent volume. If the host crashes where Vaultwarden is running on, that data is lost. We can debate cloud setups all day, and pros and cons etc... That doesn't solve your issue or my hesitations into building a maintenance feature into Vaultwarden.

Also, to be fair i think that SQLite should not be used in a HA environment. And, Vaultwarden also is not build with HA in mind at all, just a single container with a bind-mount or persistent volume.

Besides all that i quickly checked on how to even try and add this. But that isn't going to be easy. It would mean some refactoring and some way using an AtomicBool or shared Mutex or something to toggle maintenance on or off. And, then i still think users will be forced to be logged out, which is counter productive i think.

**Reply by BlackDex** [reply]
Update: Yes user will be logged-out if we can't connect to the database. And we can't just pass on read-only connections, since we do not know before the function starts if we need a read-only or also write.

Still, I do not see any issue at all by calling the backup endpoint in the admin interface to trigger a backup of the database, and run your backup pod. There will be no inconsistency with that backup db file and the state of the backup. There is a very small chance someone uploaded a file right after you run the backup command, but that is a very very small chance in my opinion.

Maybe you can change the ingress to point to a maintenance page or something else custom?
And, if you have any good other options, feel free to create a nice written PR to maybe have it included.

**Reply by 0x4E4448** [reply]
> You are still having a single point of failure when not using a persistent volume. If the host crashes where Vaultwarden is running on, that data is lost. We can debate cloud setups all day, and pros and cons etc... That doesn't solve your issue or my hesitations into building a maintenance feature into Vaultwarden.

That's incorrect. That's why you have the sidecar container that periodically copies the files to S3. Mine runs hourly, but it could be run more often. If the host fails, the pod restarts on another host and restores the data from S3.
 
> Also, to be fair i think that SQLite should not be used in a HA environment. And, Vaultwarden also is not build with HA in mind at all, just a single container with a bind-mount or persistent volume.

It's not HA in that aspect. It's a single container that can start on any one of a number of hosts.

**Reply by 0x4E4448** [reply]
> Still, I do not see any issue at all by calling the backup endpoint in the admin interface to trigger a backup of the database, and run your backup pod. There will be no inconsistency with that backup db file and the state of the backup. There is a very small chance someone uploaded a file right after you run the backup command, but that is a very very small chance in my opinion.

It's partially an issue because that requires more customization that fails if the wrong thing changes.

Is this backup file able to be used directly or does it have to be restored? As in, is it just a drop in replacement to the running database? If it's a drop in replacement, it *might* be possible without adding too much more.

**Reply by BlackDex** [reply]
It's a drop-in replacement. It's not a sql-dump or anything. It's a 100% sqlite3 file ready to be used.
You could even set the DATABASE_URL to point to that file to start using it if you want.

I'm still looking into if there is an easy way to add something like a maintenance mode. But not without causing issues to the clients as far as i can find.

**Reply by BlackDex** [reply]
@0x4E4448 I have tried several options with several HTTP response codes, but all result in a client either being logged out, or having an empty vault.
It's very difficult to only allow read-only endpoints since we do not have that configured anywhere. And changing that will take a lot of work, for which I do not see any huge improvement for Vaultwarden it self by adding this.

I think either using an external database like MySQL/MariaDB or PostgreSQL is much better suited for these kind of environments.
And if SQLite is still preferred, then use the solution mentioned above.

I don't know if you are able to easily send a custom kill command to the container? Because i was thinking about using the USR1 or USR2 kill signal to trigger a SQLite backup instead of having to use the admin endpoint which requires a login.

**Reply by 0x4E4448** [reply]
> I don't know if you are able to easily send a custom kill command to the container? Because i was thinking about using the USR1 or USR2 kill signal to trigger a SQLite backup instead of having to use the admin endpoint which requires a login.

Not from another container AFAIK and doing so would just result in it immediately restarting. Authenticating using the endpoint is less of an issue simply because the admin token is already available on the host for starting the container. It's other customizations that are the bigger issue.
